### PR TITLE
Feature/only return tags that appear more than once

### PR DESCRIPTION
**Pull request by will-sklenars**
Hey @data-doge, plz merge if you are happy with this. 
I am not sure if this is ok, declaring variables in a conditional. It seems to work as expected though. 

if (tag = Tag.find_by(word: word))
seen in dreams_controller.rb and seeds.rb

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -97,10 +97,25 @@ def retreave_tags(hash)
   end
 end
 
+# def create_themes(tag_words, dream)
+#   tag_words.each do | word |
+#     tag = Tag.find_or_create_by(word: word)
+#     Theme.create(tag: tag, dream: dream)
+#   end
+# end
+
```
Comment (data-doge) [inline db/seeds.rb]: remove?

File: app/controllers/dreams_controller.rb [hunk 2]
```diff
@@ -51,8 +51,16 @@ def retreave_tags(hash)
 
   def create_themes(tag_words, dream)
     tag_words.each do | word |
-      tag = Tag.find_or_create_by(word: word)
-      Theme.create(tag: tag, dream: dream)
+      if (tag = Tag.find_by(word: word))
+        Theme.create(tag: tag, dream: dream)
+        if (!tag.appears_more_than_once)
+          tag.appears_more_than_once = true
+          tag.save!
```
Comment (data-doge) [inline app/controllers/dreams_controller.rb]: @will-sklenars 

eyy this seems odd to me, maybe i don't understand.

my understanding is that tags should never appear more than once, so adding an `appears_more_than_once` boolean on the `Tag` model doesn't make sense to me

Comment (data-doge) [inline app/controllers/dreams_controller.rb]: @will-sklenars 

ok actually i think i understand

this boolean means like, 'there are multiple themes with this tag'

is that it?

**Comment by data-doge** [comment]
@will-sklenars i'm wondering if adding the `appears_more_than_once` boolean on the `Tag` model can be avoided.

the goal here seems to be something like:

_when i fetch the tags for a dream, i only want tags that are shared by multiple dreams_

i think that can be accomplished solely with a SQL query in `TagsController#from_dream`

i'll think about it real quick

File: app/controllers/tags_controller.rb [hunk 3]
```diff
@@ -3,6 +3,7 @@ def from_dream
     dream = Dream.find(params['dream_id'])
     themes = dream.themes
     @tags = themes.map { | theme | theme.tag }
```
Comment (data-doge) [inline app/controllers/tags_controller.rb]: it looks like you have the line `has_many :tags, through: :themes` in your `Dream` model.

this means we can replace lines 4 and 5 with: `@tags = dream.tags`. this'll be a much more efficient query than iterating through each theme to find it's tag

**Comment by data-doge** [comment]
@will-sklenars 

ok sick, i have a solution in mind -- i'll update in 20 min

File: db/seeds.rb (hunk 1 above)
Comment (will-sklenars) [inline db/seeds.rb]: yup that should be gone

File: app/controllers/dreams_controller.rb (hunk 2 above)
Comment (will-sklenars) [inline app/controllers/dreams_controller.rb]: Yea, that is correct.

File: app/controllers/tags_controller.rb (hunk 3 above)
Comment (will-sklenars) [inline app/controllers/tags_controller.rb]: mmmmmm yes that makes sense

**Comment by data-doge** [comment]
aiiight, so i'm thinking:

[1] add a `dreams_count` property to `tags`

``` ruby
class AddDreamsCountToTags < ActiveRecord::Migration
  def up
    add_column :tags, :dreams_count, :integer, default: 0, null: false
    Tag.find_each { |tag| tag.update(dreams_count: tag.dreams.count) }
  end

  def down
    remove_column :tags, :dreams_count
  end
end
```

[2] add an automatic `counter_cache` to the `Theme` model

``` ruby
class Theme < ActiveRecord::Base
  belongs_to :dream
  belongs_to :tag

  after_create :increment_counter_cache

  private
    def increment_counter_cache
      self.tag.dreams_count += 1
      self.tag.save
    end
end
```

[3] have `Tags#from_dream` return only dreams with more than one dream

``` ruby
class TagsController < ApplicationController
  def from_dream
    dream = Dream.find(params[:dream_id])
    @tags = dream.tags.where("dreams_count > ?", 1)
    render json: @tags
  end
end
```

**Comment by will-sklenars** [comment]
mmmmmm yuuup that makes sense. Then we could add a feature that when you hover over the tag hyperlink, we can have a little modal pop up that says how many dreams there are with that theme. Cool, yeah that is an elegant solution. Do you have that work in a branch to merge?

**Comment by data-doge** [comment]
naaah don't have a branch

**Comment by will-sklenars** [comment]
Cool. What do you think is the best way to implement these changes? Shall I create a new branch and put them in, or shall I leave it to you to do at a later time?

**Comment by data-doge** [comment]
yooo -- jsut got back to this, makin a branch now :yen: :yen: :yen: :yen: :yen:

**Comment by data-doge** [comment]
@will-sklenars 

kk sweet - can u review the additions i've made to this PR? merge if u want

summary of changes:
- migration to add a `dreams_count` property to `tags`
- automatic counter_cache on the `Theme` model
- `tags#for_dream` returns tags with `dreams_count > 0`
- specs for `TagsController`
- specs passing for 'DreamsController`
- refactored `DreamsController` via `after_create` hook on the `Dream` model, and moving code from `DreamsController#create` into `DreamService#create`
- removed random unused code and comments

**Comment by will-sklenars** [comment]
gr8!

**Pull request merged**